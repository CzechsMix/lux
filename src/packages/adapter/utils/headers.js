// @flow
import entries from '../../../utils/entries';
import type { ObjectMap } from '../../../interfaces';

type HandleChange = (type: 'SET' | 'DELETE', data: [string, ?string]) => void;

class Headers extends Map<string, string> {
  constructor(value: ObjectMap<string>) {
    super(entries(value));
  }

  get(key: string): void | string {
    return super.get(String(key).toLowerCase());
  }

  has(key: string): boolean {
    return super.has(String(key).toLowerCase());
  }

  set(key: string, value: string): this {
    super.set(String(key).toLowerCase(), value);
    return this;
  }

  delete(key: string): boolean {
    return super.delete(String(key).toLowerCase());
  }
}

export class RequestHeaders extends Headers {
  writable: boolean;

  constructor(value: ObjectMap<string>) {
    super(value);
    this.writable = false;
  }

  set(key: string, value: string): this {
    if (this.writable) {
      super.set(key, value);
    }
    return this;
  }

  delete(key: string): boolean {
    if (this.writable) {
      return super.delete(key);
    }
    return false;
  }
}

Object.defineProperty(RequestHeaders.prototype, 'writable', {
  value: true,
  writable: true,
  enumerable: false,
  configurable: false,
});

export class ResponseHeaders extends Headers {
  handleChange: HandleChange;

  constructor(handleChange: HandleChange) {
    super({});
    this.handleChange = handleChange;
  }

  set(key: string, value: string): this {
    this.handleChange('SET', [key, value]);
    return super.set(key, value);
  }

  delete(key: string): boolean {
    this.handleChange('DELETE', [key, null]);
    return super.delete(key);
  }
}

Object.defineProperty(ResponseHeaders.prototype, 'handleChange', {
  value: () => undefined,
  writable: true,
  enumerable: false,
  configurable: false,
});
