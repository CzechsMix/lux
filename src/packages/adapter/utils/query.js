// @flow
import entries from '../../../utils/entries';
import isObject from '../../../utils/is-object';
import type { ObjectMap } from '../../../interfaces';

const INT = /^\d+$/;
const CSV = /^(?:[\w\d-]+)(?:,[\w\d-]+)*$/;
const NULL = /^null$/;
const BOOL = /^(?:true|false)$/;
const DATE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}(Z|\+\d{4})$/;
const TRUE = /^true$/;
const DELIMITER = /[_-\s]+/;

const camelize = (source: string) => (
  source
    .split(DELIMITER)
    .reduce((result, part, idx) => {
      if (part[0]) {
        const [first] = part;

        return (
          result
          + (idx === 0 ? first.toLowerCase() : first.toUpperCase())
          + part.slice(1).toLowerCase()
        );
      }

      return result;
    }, '')
);

const fromString = (source: string): any => {
  if (INT.test(source)) {
    return Number.parseInt(source, 10);
  } else if (BOOL.test(source)) {
    return TRUE.test(source);
  } else if (NULL.test(source)) {
    return null;
  } else if (CSV.test(source)) {
    return source.split(',');
  } else if (DATE.test(source)) {
    return new Date(source);
  }
  return source;
};

export const fromObject = (source: ObjectMap<any>): ObjectMap<any> => (
  entries(source).reduce((target, [k, v]) => {
    const key = camelize(k);
    let value = v;

    if (typeof value === 'string') {
      value = fromString(value);
    } else if (isObject(value)) {
      value = fromObject(value);
    }

    if (key === 'include') {
      if (Array.isArray(value)) {
        value = value.map(camelize);
      } else if (typeof value === 'string') {
        value = [camelize(value)];
      }
    } else if (key === 'fields' && isObject(value)) {
      value = entries(value).reduce((fields, [resource, names]) => {
        // eslint-disable-next-line no-param-reassign
        fields[resource] = (
          Array.isArray(names) ? names.map(camelize) : [camelize(names)]
        );
        return fields;
      }, {});
    }

    // eslint-disable-next-line no-param-reassign
    target[key] = value;

    return target;
  }, {})
);
