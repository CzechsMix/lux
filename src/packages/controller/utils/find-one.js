// @flow
import merge from '../../../utils/merge';
// eslint-disable-next-line no-unused-vars
import type { Model, Query } from '../../database';
import type { Request } from '../../request';

import paramsToQuery from './params-to-query';

/**
 * @private
 */
export default function findOne<T: Model>(
  model: Class<T>,
  req: Request
): Query<T> {
  const params = merge(req.defaultParams, req.params);
  const { id, select, include } = paramsToQuery(model, params);

  return model.find(id)
    .select(...select)
    .include(include);
}
