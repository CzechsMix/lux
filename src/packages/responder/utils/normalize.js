// @flow
import { STATUS_CODES } from '../../../constants';
import stringify from '../../../utils/stringify';

import dataFor from './data-for';

type ResponseData = {
  data: string;
  statusCode: number;
};

/**
 * @private
 */
export default function normalize(content: any): ResponseData {
  let data;
  let statusCode;

  switch (typeof content) {
    case 'boolean':
      if (content) {
        statusCode = 204;
      } else {
        statusCode = 401;
        data = dataFor(statusCode);
      }
      break;

    case 'number':
      if (STATUS_CODES.has(content)) {
        statusCode = content;
      } else {
        statusCode = 404;
      }
      data = dataFor(statusCode);
      break;

    case 'object':
      if (content instanceof Error) {
        statusCode = Number.parseInt(content.statusCode, 10) || 500;
        data = dataFor(statusCode, content);
      } else if (!content) {
        statusCode = 404;
        data = dataFor(statusCode);
      } else {
        statusCode = 200;
        data = content;
      }
      break;

    case 'undefined':
      statusCode = 404;
      data = dataFor(statusCode);
      break;

    default:
      statusCode = 200;
      data = content;
  }

  data = stringify(data);

  return {
    data,
    statusCode,
  };
}
