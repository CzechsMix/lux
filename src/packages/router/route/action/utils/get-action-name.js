// @flow
import type { Request } from '../../../../request';

/**
 * @private
 */
export default function getActionName({
  route: {
    action
  }
}: Request): string {
  return action;
}
