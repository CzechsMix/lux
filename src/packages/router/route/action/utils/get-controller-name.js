// @flow
import type { Request } from '../../../../request';

/**
 * @private
 */
export default function getControllerName({
  route: {
    controller: {
      constructor: {
        name
      }
    }
  }
}: Request): string {
  return name;
}
