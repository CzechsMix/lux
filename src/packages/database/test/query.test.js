// @flow


import Query from '../query';
import Model from '../model';

import setType from '../../../utils/set-type';
import { getTestApp } from '../../../../test/utils/get-test-app';

describe('module "database/query"', () => {
  describe('class Query', () => {
    let Comment: Class<Model>;

    class TestModel extends Model {
      id: number;
      body: string;
      user: Class<Model>;
      tags: Array<Class<Model>>;
      title: string;
      comments: Array<Class<Model>>;
      isPublic: boolean;
      reactions: Array<Class<Model>>;
      createdAt: Date;
      updatedAt: Date;

      static tableName = 'posts';

      static belongsTo = {
        user: {
          inverse: 'posts'
        }
      };

      static hasMany = {
        comments: {
          inverse: 'post'
        },

        reactions: {
          inverse: 'post'
        },

        tags: {
          inverse: 'posts',
          through: 'categorization'
        }
      };

      static scopes = {
        isPublic() {
          return this.where({
            isPublic: true
          });
        }
      };
    }

    const assertItem = item => {
      expect(item instanceof TestModel).toBe(true);
    };

    beforeAll(async () => {
      const { store } = await getTestApp();

      Comment = store.modelFor('comment');

      await TestModel.initialize(store, () => {
        return store.connection(TestModel.tableName);
      });
    });

    describe('.from()', () => {
      let source;

      beforeAll(() => {
        source = new Query(TestModel)
          .limit(10)
          .order('title', 'DESC')
          .include(
            'user',
            'tags',
            'comments',
            'reactions'
          )
          .where({
            isPublic: true
          });
      });

      it('creates a new `Query` from a source instance of `Query`', () => {
        const result = Query.from(source);

        expect(result).not.toBe(source);
        expect(result.model).toBe(source.model);
        expect(result.isFind).toBe(source.isFind);
        expect(result.collection).toBe(source.collection);
        expect(result.shouldCount).toBe(source.shouldCount);
        expect(result.snapshots).toEqual(source.snapshots);
        expect(result.relationships).toBe(source.relationships);
      });
    });

    describe('#all()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        const result = subject.all();

        expect(result).toBe(subject);
      });

      it('does not modify #snapshots', () => {
        const { snapshots } = subject.all();

        expect(snapshots).toEqual(expect.any(Array));
        expect(snapshots).toHaveLength(0);
      });

      it('resolves with the correct array of `Model` instances', async () => {
        const result = await subject.all();

        expect(result).toEqual(expect.any(Array));
        expect(result).toHaveLength(100);
        result.forEach(assertItem);
      });
    });

    describe('#not()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        const result = subject.not({
          isPublic: true
        });

        expect(result).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.not({
          isPublic: true
        });

        expect(result.snapshots).toEqual([
          ['whereNot', { 'posts.is_public': true }]
        ]);
      });

      it('properly handles array conditions', () => {
        const result = subject.not({
          id: [1, 2, 3],
          isPublic: true
        });

        expect(result.snapshots).toEqual([
          ['whereNotIn', ['posts.id', [1, 2, 3]]],
          ['whereNot', { 'posts.is_public': true }]
        ]);
      });

      it('resolves with the correct array of `Model` instances', async () => {
        const result = await subject.not({
          isPublic: true
        });

        expect(result).toEqual(expect.any(Array));

        if (Array.isArray(result)) {
          result.forEach(item => {
            assertItem(item);
            expect(item.isPublic).toBe(false);
          });
        }
      });
    });

    describe('#find()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        const result = subject.find(1);

        expect(result).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.find(1);

        expect(result.snapshots).toEqual([
          ['where', { 'posts.id': 1 }],
          ['limit', 1]
        ]);
      });

      it('sets #isFind to `true`', () => {
        const result = subject.find(1);

        expect(result.isFind).toBe(true);
      });

      it('sets #collection to `false`', () => {
        const result = subject.find(1);

        expect(result.collection).toBe(false);
      });

      it('does not add a limit to #snapshots if #shouldCount', () => {
        subject.shouldCount = true;

        const result = subject.find(1);

        expect(result.snapshots).toEqual([
          ['where', { 'posts.id': 1 }]
        ]);
      });

      it('resolves with the correct `Model` instance', async () => {
        const result = await subject.find(1);

        expect(result).toEqual(expect.any(TestModel));
        expect(result).toEqual(
          expect.objectContaining({
            id: 1,
          })
        );
      });
    });

    describe('#page()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        const result = subject.page(2);

        expect(result).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.page(2);

        expect(result.snapshots).toEqual([
          ['limit', 25],
          ['offset', 25]
        ]);
      });

      it('does not modify #snapshots if #shouldCount', () => {
        subject.shouldCount = true;

        const result = subject.page(2);

        expect(result.snapshots).toHaveLength(0);
      });

      it('resolves with the correct array of `Model` instances', async () => {
        const result = await subject.page(2);

        expect(result).toEqual(expect.any(Array));
        expect(result).toHaveLength(25);
        result.forEach(assertItem);
      });
    });

    describe('#limit()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        const result = subject.limit(5);

        expect(result).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.limit(5);

        expect(result.snapshots).toEqual([
          ['limit', 5]
        ]);
      });

      it('does not modify #snapshots if #shouldCount', () => {
        subject.shouldCount = true;

        const result = subject.limit(5);

        expect(result.snapshots).toHaveLength(0);
      });

      it('resolves with the correct array of `Model` instances', async () => {
        const result = await subject.limit(5);

        expect(result).toEqual(expect.any(Array));
        expect(result).toHaveLength(5);
        result.forEach(assertItem);
      });
    });

    describe('#order()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        const result = subject.order('id', 'DESC');

        expect(result).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.order('id', 'DESC');

        expect(result.snapshots).toEqual([
          ['orderByRaw', 'posts.id DESC']
        ]);
      });

      it('defaults sort direction to `ASC`', () => {
        const result = subject.order('id');

        expect(result.snapshots).toEqual([
          ['orderByRaw', 'posts.id ASC']
        ]);
      });

      it('does not modify #snapshots if #shouldCount', () => {
        subject.shouldCount = true;

        const result = subject.order('id', 'DESC');

        expect(result.snapshots).toHaveLength(0);
      });

      it('resolves with the correct array of `Model` instances', async () => {
        const limit = 100;
        const result = await subject.order('id', 'DESC');

        expect(result).toEqual(expect.any(Array));
        expect(result).toHaveLength(limit);
        result.forEach(assertItem);
      });
    });

    describe('#where()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        const result = subject.where({
          isPublic: true
        });

        expect(result).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.where({
          isPublic: true
        });

        expect(result.snapshots).toEqual([
          ['where', { 'posts.is_public': true }]
        ]);
      });

      it('properly handles array conditions', () => {
        const result = subject.where({
          id: [1, 2, 3],
          isPublic: true
        });

        expect(result.snapshots).toEqual([
          ['whereIn', ['posts.id', [1, 2, 3]]],
          ['where', { 'posts.is_public': true }]
        ]);
      });

      it('resolves with the correct array of `Model` instances', async () => {
        const result = await subject.where({
          isPublic: true
        });

        expect(result).toEqual(expect.any(Array));

        if (Array.isArray(result)) {
          result.forEach(item => {
            assertItem(item);
            expect(item.isPublic).toBe(true);
          });
        }
      });
    });

    describe('#first()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        const result = subject.first();

        expect(result).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.first();

        expect(result.snapshots).toEqual([
          ['orderByRaw', 'posts.id ASC'],
          ['limit', 1]
        ]);
      });

      it('sets #collection to `false`', () => {
        const result = subject.first();

        expect(result.collection).toBe(false);
      });

      it('respects order if one already exists', () => {
        const result = subject.order('createdAt', 'DESC').first();

        expect(result.snapshots).toEqual([
          ['orderByRaw', 'posts.created_at DESC, posts.id DESC'],
          ['limit', 1]
        ]);
      });

      it('does not modify #snapshots if #shouldCount', () => {
        subject.shouldCount = true;

        const result = subject.first();

        expect(result.snapshots).toHaveLength(0);
      });

      it('resolves with the correct `Model` instance', async () => {
        const result = await subject.first();

        expect(result).toEqual(expect.any(TestModel));
        expect(result).toEqual(
          expect.objectContaining({
            id: 1,
          })
        );
      });
    });

    describe('#last()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        const result = subject.last();

        expect(result).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.last();

        expect(result.snapshots).toEqual([
          ['orderByRaw', 'posts.id DESC'],
          ['limit', 1]
        ]);
      });

      it('sets #collection to `false`', () => {
        const result = subject.last();

        expect(result.collection).toBe(false);
      });

      it('respects order if one already exists', () => {
        const result = subject.order('createdAt', 'DESC').last();

        expect(result.snapshots).toEqual([
          ['orderByRaw', 'posts.created_at DESC, posts.id DESC'],
          ['limit', 1]
        ]);
      });

      it('does not modify #snapshots if #shouldCount', () => {
        subject.shouldCount = true;

        const result = subject.last();

        expect(result.snapshots).toHaveLength(0);
      });

      it('resolves with the correct `Model` instance', async () => {
        const result = await subject.last();

        expect(result).toEqual(expect.any(TestModel));
        expect(result).toEqual(
          expect.objectContaining({
            id: 100,
          })
        );
      });
    });

    describe('#count()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        const result = subject.count();

        expect(result).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.count();

        expect(result.snapshots).toEqual([
          ['count', '* as countAll']
        ]);
      });

      it('sets #shouldCount to `true`', () => {
        const result = subject.count();

        expect(result.shouldCount).toBe(true);
      });

      it('removes all snapshots except for filter conditions', () => {
        const result = subject
          .limit(1)
          .offset(50)
          .order('createdAt')
          .where({ isPublic: true })
          .count();

        expect(result.snapshots).toMatchSnapshot();
      });

      it('resolves with the number of matching records', async () => {
        expect(await subject.count()).toBe(100);
      });
    });

    describe('#offset()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        expect(subject.offset(10)).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.offset(10);

        expect(result.snapshots).toMatchSnapshot();
      });

      it('does not modify #snapshots if #shouldCount', () => {
        subject.shouldCount = true;

        const result = subject.offset(10);

        expect(result.snapshots).toMatchSnapshot();
      });

      it('resolves with the correct array of `Model` instances', async () => {
        const result = await subject.offset(10);

        expect(result).toEqual(expect.any(Array));
        expect(result).toHaveLength(90);
        result.forEach(assertItem);
      });
    });

    describe('#select()', () => {
      let subject;
      const attrs = ['id', 'title', 'createdAt'];

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        expect(subject.select(...attrs)).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.select(...attrs);

        expect(result.snapshots).toMatchSnapshot();
      });

      it('resolves with the correct array of `Model` instances', async () => {
        const result = await subject.select(...attrs);

        expect(result).toEqual(expect.any(Array));

        if (Array.isArray(result)) {
          result.forEach(item => {
            assertItem(item);
            expect(Object.keys(item.rawColumnData)).toEqual(attrs);;
          });
        }
      });
    });

    describe('#distinct()', () => {
      let subject;
      const attrs = ['title'];

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        expect(subject.distinct(...attrs)).toBe(subject);
      });

      it('properly modifies #snapshots', () => {
        const result = subject.distinct(...attrs);

        expect(result.snapshots).toMatchSnapshot();
      });

      it('resolves with the correct array of `Model` instances', async () => {
        const result = await subject.distinct(...attrs);

        expect(result).toEqual(expect.any(Array));

        if (Array.isArray(result)) {
          result.forEach(item => {
            assertItem(item);
            expect(Object.keys(item.rawColumnData)).toEqual(attrs);;
          });
        }
      });
    });

    describe('#include()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        expect(subject.include('user', 'comments')).toBe(subject);
      });

      it('properly modifies #snapshots when using an array of strings', () => {
        const result = subject.include('user', 'comments');

        expect([
          result.snapshots,
          result.relationships,
        ]).toMatchSnapshot();
      });

      it('properly modifies #snapshots when using an object', () => {
        const result = subject.include({
          user: [
            'id',
            'name'
          ],
          comments: [
            'id',
            'name',
            'edited',
            'updatedAt'
          ]
        });

        expect([
          result.snapshots,
          result.relationships,
        ]).toMatchSnapshot();
      });

      it('resolves with the correct array of `Model` instances', async () => {
        const result = await subject.include('user', 'comments');

        expect(result).toMatchSnapshot();
      });
    });

    describe('#scope()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('can be chained to other query methods', () => {
        const result = subject
          .isPublic()
          .select('id', 'title')
          .limit(10);

        expect(result.snapshots).toMatchSnapshot();
      });

      it('can be chained from other query methods', () => {
        const result = subject
          .all()
          .select('id', 'title')
          .limit(10)
          .isPublic();

        expect(result.snapshots).toMatchSnapshot();
      });
    });

    describe('#unscope()', () => {
      let subject;

      beforeEach(() => {
        subject = new Query(TestModel);
      });

      it('returns `this`', () => {
        expect(subject.isPublic().unscope('isPublic')).toBe(subject);
      });

      it('removes a named scope from #snapshots', () => {
        const result = subject
          .select('id', 'title')
          .isPublic()
          .limit(10)
          .unscope('isPublic');

        expect(result.snapshots).toMatchSnapshot();
      });
    });
  });
});
