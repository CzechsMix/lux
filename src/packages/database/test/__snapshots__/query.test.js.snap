exports[`module "database/query" class Query #all() does not modify #snapshots 1`] = `Array []`;

exports[`module "database/query" class Query #all() resolves with the correct array of \`Model\` instances 1`] = `
Array [
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
  Test {},
]
`;

exports[`module "database/query" class Query #count() properly modifies #snapshots 1`] = `
Array [
  Array [
    "count",
    "* as countAll",
  ],
]
`;

exports[`module "database/query" class Query #count() removes all snapshots except for filter conditions 1`] = `
Array [
  Array [
    "count",
    "* as countAll",
  ],
]
`;

exports[`module "database/query" class Query #distinct() properly modifies #snapshots 1`] = `
Array [
  Array [
    "distinct",
    Array [
      "posts.title AS title",
    ],
  ],
  Array [
    "select",
    Array [],
  ],
]
`;

exports[`module "database/query" class Query #include() properly modifies #snapshots when using an array of strings 1`] = `
Array [
  Array [
    Array [
      "includeSelect",
      Array [
        "users.id AS user.id",
        "users.name AS user.name",
        "users.email AS user.email",
        "users.password AS user.password",
        "users.created_at AS user.createdAt",
        "users.updated_at AS user.updatedAt",
      ],
    ],
    Array [
      "leftOuterJoin",
      Array [
        "users",
        "posts.user_id",
        "=",
        "users.id",
      ],
    ],
  ],
  Object {
    "comments": Object {
      "attrs": Array [
        "id",
        "message",
        "edited",
        "userId",
        "postId",
        "createdAt",
        "updatedAt",
        "postId",
      ],
      "foreignKey": "post_id",
      "model": [Function],
      "through": undefined,
      "type": "hasMany",
    },
  },
]
`;

exports[`module "database/query" class Query #include() properly modifies #snapshots when using an object 1`] = `
Array [
  Array [
    Array [
      "includeSelect",
      Array [
        "users.id AS user.id",
        "users.name AS user.name",
      ],
    ],
    Array [
      "leftOuterJoin",
      Array [
        "users",
        "posts.user_id",
        "=",
        "users.id",
      ],
    ],
  ],
  Object {
    "comments": Object {
      "attrs": Array [
        "id",
        "name",
        "edited",
        "updatedAt",
        "postId",
      ],
      "foreignKey": "post_id",
      "model": [Function],
      "through": undefined,
      "type": "hasMany",
    },
  },
]
`;

exports[`module "database/query" class Query #include() resolves with the correct array of \`Model\` instances 1`] = `
Array [
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
  TestModel {},
]
`;

exports[`module "database/query" class Query #last() properly modifies #snapshots 1`] = `
Array [
  Array [
    "orderByRaw",
    "tests.id DESC",
  ],
  Array [
    "limit",
    1,
  ],
]
`;

exports[`module "database/query" class Query #last() respects order if one already exists 1`] = `
Array [
  Array [
    "orderByRaw",
    "tests.created_at DESC, tests.id DESC",
  ],
  Array [
    "limit",
    1,
  ],
]
`;

exports[`module "database/query" class Query #offset() does not modify #snapshots if #shouldCount 1`] = `Array []`;

exports[`module "database/query" class Query #offset() properly modifies #snapshots 1`] = `
Array [
  Array [
    "offset",
    10,
  ],
]
`;

exports[`module "database/query" class Query #order() defaults sort direction to \`ASC\` 1`] = `
Array [
  Array [
    "orderByRaw",
    "tests.id ASC",
  ],
]
`;

exports[`module "database/query" class Query #order() properly modifies #snapshots 1`] = `
Array [
  Array [
    "orderByRaw",
    "tests.id DESC",
  ],
]
`;

exports[`module "database/query" class Query #order() resolves with the correct array of \`Model\` instances 1`] = `
Array [
  100,
  99,
  98,
  97,
  96,
  95,
  94,
  93,
  92,
  91,
  90,
  89,
  88,
  87,
  86,
  85,
  84,
  83,
  82,
  81,
  80,
  79,
  78,
  77,
  76,
  75,
  74,
  73,
  72,
  71,
  70,
  69,
  68,
  67,
  66,
  65,
  64,
  63,
  62,
  61,
  60,
  59,
  58,
  57,
  56,
  55,
  54,
  53,
  52,
  51,
  50,
  49,
  48,
  47,
  46,
  45,
  44,
  43,
  42,
  41,
  40,
  39,
  38,
  37,
  36,
  35,
  34,
  33,
  32,
  31,
  30,
  29,
  28,
  27,
  26,
  25,
  24,
  23,
  22,
  21,
  20,
  19,
  18,
  17,
  16,
  15,
  14,
  13,
  12,
  11,
  10,
  9,
  8,
  7,
  6,
  5,
  4,
  3,
  2,
  1,
]
`;

exports[`module "database/query" class Query #scope() can be chained from other query methods 1`] = `
Array [
  Array [
    "select",
    Array [
      "posts.id AS id",
      "posts.title AS title",
    ],
  ],
  Array [
    "limit",
    10,
  ],
  Array [
    "where",
    Object {
      "posts.is_public": true,
    },
    "isPublic",
  ],
]
`;

exports[`module "database/query" class Query #scope() can be chained to other query methods 1`] = `
Array [
  Array [
    "where",
    Object {
      "posts.is_public": true,
    },
    "isPublic",
  ],
  Array [
    "select",
    Array [
      "posts.id AS id",
      "posts.title AS title",
    ],
  ],
  Array [
    "limit",
    10,
  ],
]
`;

exports[`module "database/query" class Query #select() properly modifies #snapshots 1`] = `
Array [
  Array [
    "select",
    Array [
      "posts.id AS id",
      "posts.title AS title",
      "posts.created_at AS createdAt",
    ],
  ],
]
`;

exports[`module "database/query" class Query #unscope() removes a named scope from #snapshots 1`] = `
Array [
  Array [
    "select",
    Array [
      "posts.id AS id",
      "posts.title AS title",
    ],
  ],
  Array [
    "limit",
    10,
  ],
]
`;
